### ISSUE TITLE
The relationship between formatters and dplyr/tibble is confusing

### ISSUE THREAD

**Issue by pawelru** [insightsengineering/formatters#165]
Hello. I am currently working on the failing minimal integration tests on `rlistings` where we have many tests that covers expectations on column labels:
https://github.com/insightsengineering/rlistings/blob/main/tests/testthat/test-listings.R#L6
https://github.com/insightsengineering/rlistings/blob/3cad52ad370e09522c7173bd8b85d5481a883584/tests/testthat/test-print.R#L5-L23
https://github.com/insightsengineering/rlistings/blob/3cad52ad370e09522c7173bd8b85d5481a883584/tests/testthat/test-print.R#L30-L76
I believe that such tests should live in formatters and below findings should came from formatters failuers.

To the point:
- using the newest `dplyr` -> `install.packages("dplyr", quiet = TRUE, repos = "https://cloud.r-project.org")`
``` r
library(formatters)
anl <- ex_adsl[1:10, c("USUBJID", "ARM", "BMRKR1")]
var_labels(anl)["BMRKR1"]
#> BMRKR1 
#>     NA
anl <- var_relabel(anl,
  USUBJID = "Unique\nSubject\nIdentifier",
  ARM = "Description\nOf\nPlanned Arm"
)
var_labels(anl)["BMRKR1"]
#> BMRKR1 
#>     NA
sessionInfo()
#> R version 4.2.1 (2022-06-23)
#> Platform: x86_64-apple-darwin17.0 (64-bit)
#> Running under: macOS Big Sur ... 10.16
#> 
#> Matrix products: default
#> BLAS:   /Library/Frameworks/R.framework/Versions/4.2/Resources/lib/libRblas.0.dylib
#> LAPACK: /Library/Frameworks/R.framework/Versions/4.2/Resources/lib/libRlapack.dylib
#> 
#> locale:
#> [1] en_US.UTF-8/en_US.UTF-8/en_US.UTF-8/C/en_US.UTF-8/en_US.UTF-8
#> 
#> attached base packages:
#> [1] stats     graphics  grDevices utils     datasets  methods   base     
#> 
#> other attached packages:
#> [1] formatters_0.5.0
#> 
#> loaded via a namespace (and not attached):
#>  [1] digest_0.6.31     withr_2.5.0       R.methodsS3_1.8.2 grid_4.2.1       
#>  [5] backports_1.4.1   lifecycle_1.0.3   magrittr_2.0.3    reprex_2.0.2     
#>  [9] evaluate_0.21     rlang_1.1.1       cli_3.6.1         fs_1.6.2         
#> [13] R.utils_2.12.2    R.oo_1.25.0       checkmate_2.0.0   vctrs_0.6.2      
#> [17] rmarkdown_2.20    styler_1.10.1     tools_4.2.1       R.cache_0.16.0   
#> [21] glue_1.6.2        purrr_1.0.1       xfun_0.37         yaml_2.3.7       
#> [25] fastmap_1.1.1     compiler_4.2.1    htmltools_0.5.5   knitr_1.42
```

<sup>Created on 2023-06-12 with [reprex v2.0.2](https://reprex.tidyverse.org)</sup>

``` r
library(formatters)
library(dplyr)
#> 
#> Attaching package: 'dplyr'
#> The following objects are masked from 'package:stats':
#> 
#>     filter, lag
#> The following objects are masked from 'package:base':
#> 
#>     intersect, setdiff, setequal, union
anl <- ex_adsl[1:10, c("USUBJID", "ARM", "BMRKR1")]
var_labels(anl)["BMRKR1"]
#>                        BMRKR1 
#> "Continous Level Biomarker 1"
anl <- var_relabel(anl,
  USUBJID = "Unique\nSubject\nIdentifier",
  ARM = "Description\nOf\nPlanned Arm"
)
var_labels(anl)["BMRKR1"]
#>                        BMRKR1 
#> "Continous Level Biomarker 1"
sessionInfo()
#> R version 4.2.1 (2022-06-23)
#> Platform: x86_64-apple-darwin17.0 (64-bit)
#> Running under: macOS Big Sur ... 10.16
#> 
#> Matrix products: default
#> BLAS:   /Library/Frameworks/R.framework/Versions/4.2/Resources/lib/libRblas.0.dylib
#> LAPACK: /Library/Frameworks/R.framework/Versions/4.2/Resources/lib/libRlapack.dylib
#> 
#> locale:
#> [1] en_US.UTF-8/en_US.UTF-8/en_US.UTF-8/C/en_US.UTF-8/en_US.UTF-8
#> 
#> attached base packages:
#> [1] stats     graphics  grDevices utils     datasets  methods   base     
#> 
#> other attached packages:
#> [1] dplyr_1.1.2      formatters_0.5.0
#> 
#> loaded via a namespace (and not attached):
#>  [1] knitr_1.42        magrittr_2.0.3    tidyselect_1.2.0  R.cache_0.16.0   
#>  [5] R6_2.5.1          rlang_1.1.1       fastmap_1.1.1     fansi_1.0.4      
#>  [9] styler_1.10.1     tools_4.2.1       grid_4.2.1        checkmate_2.0.0  
#> [13] xfun_0.37         R.oo_1.25.0       utf8_1.2.3        cli_3.6.1        
#> [17] withr_2.5.0       htmltools_0.5.5   yaml_2.3.7        digest_0.6.31    
#> [21] tibble_3.2.1      lifecycle_1.0.3   purrr_1.0.1       vctrs_0.6.2      
#> [25] R.utils_2.12.2    fs_1.6.2          glue_1.6.2        evaluate_0.21    
#> [29] rmarkdown_2.20    reprex_2.0.2      pillar_1.9.0      compiler_4.2.1   
#> [33] generics_0.1.3    backports_1.4.1   R.methodsS3_1.8.2 pkgconfig_2.0.3
```

<sup>Created on 2023-06-12 with [reprex v2.0.2](https://reprex.tidyverse.org)</sup>

- using older `dplyr@1.0.0` + `tibble@2.1.3` (!! note: this is about those two packages: I checked different versions of dplyr while keeping the newest `tibble` and there were no differences) -> `devtools::install_version("dplyr", "1.0.0", quiet = TRUE, repos = "https://cloud.r-project.org"); devtools::install_version("tibble", "2.1.3", quiet = TRUE, repos = "https://cloud.r-project.org")`

``` r
library(formatters)
anl <- ex_adsl[1:10, c("USUBJID", "ARM", "BMRKR1")]
var_labels(anl)["BMRKR1"]
#> BMRKR1 
#>     NA
anl <- var_relabel(anl,
  USUBJID = "Unique\nSubject\nIdentifier",
  ARM = "Description\nOf\nPlanned Arm"
)
var_labels(anl)["BMRKR1"]
#> BMRKR1 
#>     NA
sessionInfo()
#> R version 4.2.1 (2022-06-23)
#> Platform: x86_64-apple-darwin17.0 (64-bit)
#> Running under: macOS Big Sur ... 10.16
#> 
#> Matrix products: default
#> BLAS:   /Library/Frameworks/R.framework/Versions/4.2/Resources/lib/libRblas.0.dylib
#> LAPACK: /Library/Frameworks/R.framework/Versions/4.2/Resources/lib/libRlapack.dylib
#> 
#> locale:
#> [1] en_US.UTF-8/en_US.UTF-8/en_US.UTF-8/C/en_US.UTF-8/en_US.UTF-8
#> 
#> attached base packages:
#> [1] stats     graphics  grDevices utils     datasets  methods   base     
#> 
#> other attached packages:
#> [1] formatters_0.5.0
#> 
#> loaded via a namespace (and not attached):
#>  [1] digest_0.6.31     withr_2.5.0       R.methodsS3_1.8.2 grid_4.2.1       
#>  [5] backports_1.4.1   lifecycle_1.0.3   magrittr_2.0.3    reprex_2.0.2     
#>  [9] evaluate_0.21     rlang_1.1.1       cli_3.6.1         fs_1.6.2         
#> [13] R.utils_2.12.2    R.oo_1.25.0       checkmate_2.0.0   vctrs_0.6.2      
#> [17] rmarkdown_2.20    styler_1.10.1     tools_4.2.1       R.cache_0.16.0   
#> [21] glue_1.6.2        purrr_1.0.1       xfun_0.37         yaml_2.3.7       
#> [25] fastmap_1.1.1     compiler_4.2.1    htmltools_0.5.5   knitr_1.42
```

<sup>Created on 2023-06-12 with [reprex v2.0.2](https://reprex.tidyverse.org)</sup>

``` r
library(formatters)
library(dplyr)
#> Registered S3 methods overwritten by 'tibble':
#>   method     from  
#>   format.tbl pillar
#>   print.tbl  pillar
#> Warning: replacing previous import 'vctrs::data_frame' by 'tibble::data_frame'
#> when loading 'dplyr'
#> 
#> Attaching package: 'dplyr'
#> The following objects are masked from 'package:stats':
#> 
#>     filter, lag
#> The following objects are masked from 'package:base':
#> 
#>     intersect, setdiff, setequal, union
anl <- ex_adsl[1:10, c("USUBJID", "ARM", "BMRKR1")]
var_labels(anl)["BMRKR1"]
#> BMRKR1 
#>     NA
anl <- var_relabel(anl,
  USUBJID = "Unique\nSubject\nIdentifier",
  ARM = "Description\nOf\nPlanned Arm"
)
var_labels(anl)["BMRKR1"]
#> BMRKR1 
#>     NA
sessionInfo()
#> R version 4.2.1 (2022-06-23)
#> Platform: x86_64-apple-darwin17.0 (64-bit)
#> Running under: macOS Big Sur ... 10.16
#> 
#> Matrix products: default
#> BLAS:   /Library/Frameworks/R.framework/Versions/4.2/Resources/lib/libRblas.0.dylib
#> LAPACK: /Library/Frameworks/R.framework/Versions/4.2/Resources/lib/libRlapack.dylib
#> 
#> locale:
#> [1] en_US.UTF-8/en_US.UTF-8/en_US.UTF-8/C/en_US.UTF-8/en_US.UTF-8
#> 
#> attached base packages:
#> [1] stats     graphics  grDevices utils     datasets  methods   base     
#> 
#> other attached packages:
#> [1] dplyr_1.0.0      formatters_0.5.0
#> 
#> loaded via a namespace (and not attached):
#>  [1] knitr_1.42        magrittr_2.0.3    tidyselect_1.2.0  R.cache_0.16.0   
#>  [5] R6_2.5.1          rlang_1.1.1       fastmap_1.1.1     fansi_1.0.4      
#>  [9] styler_1.10.1     tools_4.2.1       grid_4.2.1        checkmate_2.0.0  
#> [13] xfun_0.37         R.oo_1.25.0       utf8_1.2.3        cli_3.6.1        
#> [17] withr_2.5.0       htmltools_0.5.5   yaml_2.3.7        digest_0.6.31    
#> [21] tibble_2.1.3      lifecycle_1.0.3   purrr_1.0.1       vctrs_0.6.2      
#> [25] R.utils_2.12.2    fs_1.6.2          glue_1.6.2        evaluate_0.21    
#> [29] rmarkdown_2.20    reprex_2.0.2      pillar_1.9.0      compiler_4.2.1   
#> [33] generics_0.1.3    backports_1.4.1   R.methodsS3_1.8.2 pkgconfig_2.0.3
```

<sup>Created on 2023-06-12 with [reprex v2.0.2](https://reprex.tidyverse.org)</sup>

So to sum up:
- new dplyr & tibble: empty labels -> library(dplyr) -> non-empty labels
- old dplyr & tibble: empty labels -> library(dplyr) -> empty labels

(There are many versions between dplyr@1.1.2 + tibble@3.1.2 and dplyr@1.0.0 + tibble@2.1.3 - I haven't yet tracked down from from where we observe change in behaviour.)

From what I can see in the tests in `rlistings` it seems that we are expecting labels, i.e. more confusing scenario where labels appeared after `library(dplyr)`.

My goal for this task is to:
- define the target behaviour
- specify relationships in the DESCRIPTION file of `formatters` OR make the functionality totally independent

**Comment by gmbecker** [comment]
So labels have "never" worked consistently with non-tibble data.frames, I think. At least not safely. Honestly the whole labels live in attributes thing was always a terrible idea and is in formatters at all over my objections (and via code I didn't write). attributes are notoriously fickle and you basically should never use them for anything you care about.

That said, I don't think it has anything to do with dplyr.

I see the bugged (ie "old version") behavior with dplyr 1.0.0 and tibble 2.1.3 and I see the correct (ie "new version") behavior with dplyr 1.0.0 and tibble 3.0.0 (the next version after 2.1.3). dplyr 1.1.1 (the version I had this morning and likely be installing again) depends on tibble >=3.2.0 itself, so we can't test old-tibble new dplyr.

So formatters depends on tibble 3.0.0 and I don't see any evidence of a major versioned dep on dplyr from this issue. I will add that to formatters.

**Comment by pawelru** [comment]
> I see the bugged (ie "old version") behavior with dplyr 1.0.0 and tibble 2.1.3 and I see the correct (ie "new version") behavior with dplyr 1.0.0 and tibble 3.0.0 (the next version after 2.1.3).

Just to confirm: which behaviour is bugged to you and which is correct?

**Comment by gmbecker** [comment]
The documentation says "  Relabel a subset of the variables" (in point of fact, thats _all_ it says...)

That suggests to me that you would expect that you're relabeling some of the variables, and not changing the labels on the others. So for that code snippet (the one where you are relabeling anl and assinging it back into anl, at least) should give the old label ("Biomarker <whatever whatever>"), not NA

Also as additional confirmation, tibble 3.0.0 and dplyr 0.8.1 (pretty much the earliest I can easily install on my system) has the same behavior as tibble 3.0.0 and dplyr 1.1.1, so its a pure tibble problem, dplyr is not related.

**Comment by pawelru** [comment]
> the one where you are relabeling anl and assinging it back into anl, at least

I messed up the variables and repeated the exercise (edited my post) but I got the same result.

I have also included print labels _before_ `var_relabel` and I noticed something interesting - using the newest `tibble`, labels persists when subsetting rows and columns. That's not true using older version. Then it looks like `var_relabels` works correct (i.e. return non-empty if got non-empty; return empty if empty) - it is the input that changes. Very interesting. Then I guess there is nothing to be done here and I need to incorporate that requirement into `rlistings` to require ~the newest~ _higher_ version of `tibble` to match the expected test values.

**Comment by pawelru** [comment]
OK it seems that I got confirmation of the above:

- `tibble@2.1.3` drops the labels when subsetting

``` r
remove.packages("tibble")
#> Removing package from '/Library/Frameworks/R.framework/Versions/4.2/Resources/library'
#> (as 'lib' is unspecified)
remove.packages("dplyr")
#> Removing package from '/Library/Frameworks/R.framework/Versions/4.2/Resources/library'
#> (as 'lib' is unspecified)
devtools::install_version("dplyr", "1.0.0", quiet = TRUE, repos = "https://cloud.r-project.org")
#> Installing 1 packages: tibble
devtools::install_version("tibble", "2.1.3", quiet = TRUE, repos = "https://cloud.r-project.org")
#devtools::install_version("tibble", "3.0.0", quiet = TRUE, repos = "https://cloud.r-project.org")
###############
library(formatters)
var_labels(ex_adsl)["BMRKR1"]
#>                        BMRKR1 
#> "Continous Level Biomarker 1"
var_labels(ex_adsl[1:10, c("USUBJID", "ARM", "BMRKR1")])["BMRKR1"]
#> BMRKR1 
#>     NA
library(tibble)
#> Registered S3 methods overwritten by 'tibble':
#>   method     from  
#>   format.tbl pillar
#>   print.tbl  pillar
var_labels(ex_adsl[1:10, c("USUBJID", "ARM", "BMRKR1")])["BMRKR1"]
#> BMRKR1 
#>     NA
```

<sup>Created on 2023-06-13 with [reprex v2.0.2](https://reprex.tidyverse.org)</sup>

- `tibble@3.0.0` keeps the labels when subsetting

``` r
remove.packages("tibble")
#> Removing package from '/Library/Frameworks/R.framework/Versions/4.2/Resources/library'
#> (as 'lib' is unspecified)
remove.packages("dplyr")
#> Removing package from '/Library/Frameworks/R.framework/Versions/4.2/Resources/library'
#> (as 'lib' is unspecified)
devtools::install_version("dplyr", "1.0.0", quiet = TRUE, repos = "https://cloud.r-project.org")
#> Installing 1 packages: tibble
#devtools::install_version("tibble", "2.1.3", quiet = TRUE, repos = "https://cloud.r-project.org")
devtools::install_version("tibble", "3.0.0", quiet = TRUE, repos = "https://cloud.r-project.org")
###############
library(formatters)
var_labels(ex_adsl)["BMRKR1"]
#>                        BMRKR1 
#> "Continous Level Biomarker 1"
var_labels(ex_adsl[1:10, c("USUBJID", "ARM", "BMRKR1")])["BMRKR1"]
#> BMRKR1 
#>     NA
library(tibble)
#> Registered S3 methods overwritten by 'tibble':
#>   method     from  
#>   format.tbl pillar
#>   print.tbl  pillar
var_labels(ex_adsl[1:10, c("USUBJID", "ARM", "BMRKR1")])["BMRKR1"]
#>                        BMRKR1 
#> "Continous Level Biomarker 1"
```

<sup>Created on 2023-06-13 with [reprex v2.0.2](https://reprex.tidyverse.org)</sup>


From that perspective, `var_labels` behaviour is correct. The conclusion is that if there is an expectation of non-relabeled column label presence then that indicates requirement of `tibble >= 3.0.0`. 
Closing this issue.
